src/libcrun/handlers/krun.c

```diff
@@ -293,6 +295,58 @@ libkrun_unload (void *cookie, libcrun_error_t *err arg_unused)
   return 0;
 }
 
+static runtime_spec_schema_defs_linux_device_cgroup *
+make_oci_spec_dev (const char *type, dev_t device, bool allow, const char *access)
+{
+  runtime_spec_schema_defs_linux_device_cgroup *dev = xmalloc0 (sizeof (*dev));
+
+  dev->allow = allow;
+  dev->allow_present = 1;
+
+  dev->type = xstrdup (type);
+
+  dev->major = major (device);
+  dev->major_present = 1;
+
+  dev->minor = minor (device);
+  dev->minor_present = 1;
+
+  dev->access = xstrdup (access);
+
+  return dev;
+}
+
+static int
+libkrun_modify_oci_configuration (void *cookie, libcrun_context_t *context,
+                                  runtime_spec_schema_config_schema *def,
+                                  libcrun_error_t *err)
+{
+  const size_t device_size = sizeof (runtime_spec_schema_defs_linux_device_cgroup);
+  struct stat st;
+  size_t len;
+  int ret;
+
+  if (def->linux == NULL || def->linux->resources == NULL
+      || def->linux->resources->devices == NULL)
+    return 0;
+
+  /* Always allow the /dev/kvm device.  */
+
+  ret = stat ("/dev/kvm", &st);
+  if (UNLIKELY (ret < 0))
+    return crun_make_error (err, errno, "stat `/dev/kvm`");
+
+  len = def->linux->resources->devices_len;
+  def->linux->resources->devices = xrealloc (def->linux->resources->devices,
+                                             device_size * (len + 2));
+
+  def->linux->resources->devices[len] = make_oci_spec_dev ("a", st.st_rdev, true, "rwm");
```

Comment:
Seems like adding the device to `def->linux->resources->devices` gets it allowed by cgroup, but to have it created by `libcrun_create_dev` needs to be added to `def->linux->devices` too. Alternatively we could keep the node creation logic at `libkrun_configure_container()`, but seems to be like leaving this to be dealt by the common code at `src/libcrun/linux.c` is more future-proof.

Reply:
oh let me drop this patch for now